### PR TITLE
Add MetricsContext to enable metrics tracking with initial S3FileIO implementation

### PR DESCRIPTION
**Pull request by danielcweeks**
This PR adds a basic interface for reporting FileIO metrics and a shim implementation for reporting IO metrics via the Hadoop `FileSystem.Statistics` to be compatible with how Spark/Flink gather and report task metrics.

### CODE REVIEW

**Comment by danielcweeks** [comment]
@RussellSpitzer @jackye1995 and @szehon-ho You may be interested in this since I think you've run into issues with the FileIO implementation not reporting metrics back through Spark/Trino/Flink.  This is still work-in-progress and i just implemented to the read path to get feedback on this as an approach, but I was able to verify that the read/write metrics do appear in the Spark UI.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Optional interface for FileIO implementations to report metrics related
+ * to read/write operations.
+ */
+public interface FileIOMetrics extends Serializable {
+  void initialize(Map<String,String> properties);
+  void incrementBytesRead(long bytesRead);
```
Comment (danielcweeks) [inline api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:30]: The more I think about this interface, the less I like the parallels with the Hadoop FileSystem statistics because it is very limited.

I think what we really want is a way to provide arbitrary IO specific metrics, not just bytes/ops (e.g. s3 request count, parts uploaded, request latency).  So, I think we want to pivot this interface to accept more generic metrics:

```
void increment(<Metric>)
void decrement(<Metric>) ?? Not sure if there is a use case for this
```

Where a metric is a `name/value` pair.  In the hadoop implementation, we can still map those to the statistics, but overall, it would provide more flexible metric reporting.

Thoughts?

Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:30]: I strongly agree that we should be thinking more high level, with `Counter` types that can operate on metrics with a label set (like increment(s3ResdRequestCounts).

A good example of this is Prometheus, which has counters / gauges / histograms with arbitrary labels. So we could count something like S3 request with dimensions of request type (PUT, DELETE, etc) and possibly even more dimensions like bucket.

One decent example in my opinion is Flink’s metric system that allows some of the work to be offloaded to a library (in their case, Dropwizard for histograms).

The metrics of the Hadoop FileSystem are not really what users of object storage like s3 are commonly interested in. They want to see S3 request rates and other things that impact their bill.

So a more generic way of tracking metrics would be better.

Then, engine by engine, we can provide instances / shims for those metrics to report them as is supported. But there are ways to still export the metrics in commonly used formats, like the Prometheus time series metrics format, in many engines (such as Flink - which I’ve done in production using custom metrics so I know it is possible).

Given there’s strong reason to believe that we could export our custom metrics in at least one commonly used metrics format (Prometheus for sure - and likely more), in one commonly used engine (Apache Flink) if not more, I’d vote for a higher level abstraction that we then use to implement Hadoop FileSystem metrics as needed.

+1 to more generic Counters for tracking more than just what’s supported today by Hadoop’s interfaces.

Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:30]: The Prometheus label based metrics are the best that I know of, but a decent interface  for declaring single metrics that are easily portable across metrics systems is in Apache Flink.

**Comment by kbendick** [comment]
Thank you @danielcweeks for getting the ball rolling  on this.

The Spark S3 metrics are a big concern to many people I talk to, so even supporting just Hadoop FileSystem metrics for S3FileIO initially would be a huge win.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 1 above)
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:30]: I think Iceberg would also benefit from having real observability metrics that are exposed via Counters, Timers, and such. This would allow having actual Dashboards to represent those metrics.

Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java]: Yes, I agree that we want something more like micrometer, but I don't want to tie directly.  I just pushed an updated example that I think would tie in well with a micrometer implementation.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Get a named counter of a specific type.  Metric implementations may impose
+   * restrictions on what types are supported for specific counters.
+   *
+   * @param name name of the metric
+   * @param type numeric type of the counter value
+   * @return a counter implementation
+   */
+  <T> Counter<T> counter(String name, Class<T> type);
```
Comment (danielcweeks) [inline api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:48]: For now, I'm only providing a Counter interface due to the first implementation being Hadoop metrics, which don't require other meters (e.g. Timer, Gauge).  However, this could be expanded upon similarly if those use cases emerge.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Optional interface for FileIO implementations to report metrics related
+ * to read/write operations.
+ */
+public interface FileIOMetrics extends Serializable {
+  String READ_BYTES = "read.bytes";
```
Comment (nastra) [inline api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:29]: I think one of the challenges might become the metric names and their usage with different systems. Prometheus for example uses suffixes like `_seconds` / `_bytes` / `_total` (https://prometheus.io/docs/practices/naming/). One option would we to have a metric name converter to match whatever the target metric system expects it to be.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 2 above)
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:48]: I guess for a super simple representation of metrics this might be enough. Once requirements get a little bit more sophisticated (by wanting to have Tag support for metrics), then it's possible that maintaining an own metric implementation gets more difficult.

Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:48]: I think the implementation could still add tags based on the metric name and property based initialization.  I don't want to overcomplicate or be too opinionated about the implementation.  To your point, we could always expand upon the implementation, but the first use case (hadoop fs reporting) is really limited in scope.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 3 above)
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:29]: correct, that's where I was trying to get at. We would probably need some kind of naming conversion to happen to map to whatever system we'd want the metrics to be mapped to

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 2 above)
Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:48]: I agree on both accounts.

Tags can be really useful and the biggest complaint by Flink users when I administered Flink clusters was not being able to used tagged metrics (though not all metrics systems support them directly - many still use the one metric maps to one key value pair basically). However, we eventually were able to use tagged metrics directly via the prometheus Java library directly within Flink by some configuration that Flink's metric system provided.

I imagine Micrometer might have also been able to do the same (if it was exporting metrics in a tagged format).

I agree that for the first use case, just a simple Counter is sufficient and given how limited the hadoop fs metrics reporting is in scope (and how requested those metrics are from S3FileIO), it would be better to not complicate things initially.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 3 above)
Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:29]: I'd definitely encourage you to take a look at the Flink metric implementation. It worked well for me to be able to use their metrics, plus my own metrics via the Prometheus metrics library, and merge them together. And it provides shims to various metric formats and is used for the same purpose (hadoop fs metrics): https://github.com/apache/flink/tree/master/flink-metrics

But I still agree that the simplest approach should be chosen initially to get the metrics working for S3FileIO. It wouldn't be hard to adapt them later and it's a highly requested feature.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 4]
```diff
@@ -81,6 +89,8 @@ public int read() throws IOException {
 
     pos += 1;
     next += 1;
+    readBytes.increment(1L);
```
Comment (RussellSpitzer) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:92]: Do we want to update the metrics every time here? This feels like it may be expensive although I obviously would want to check benchmarks to be sure. I think a lot of systems tend to just update metrics every X operations and then do a final update in the "close" method

**Review by RussellSpitzer — COMMENTED** [review]
So the idea would be every engine would pick via their Catalog Configuration which metrics implementation their FileIO should use. This implementation should be compatible with how the engine reports statistics or could be to an outside system if you don't want metrics for your engine. 

I think this seems pretty good to me overall

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java (hunk 4 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:92]: Yeah, this method shouldn't really be called at all (if you go back to the original implementation we were considering not even including it).  This is more for completeness of the InputStream api as opposed to trying to be performant.

tl;dr don't use this method :)

**Comment by danielcweeks** [comment]
@izchen Just wanted to bring this to your attention because #4020 was attempting to address a similar issue.   I think that issue only addressed the write side, but this would allow for better reporting for read and write metrics.

**Review by szehon-ho — COMMENTED** [review]
Yea it looks good to me too, definitely have heard some asks for it!  Two overall questions:

1.  How about the implementation for S3FileOutputStream to track write_bytes, write_records?  
2.  Not sure what is the right way, but if I am implementing a FileIOMetrics for S3FileIO, the FileIOMetrics interface is not descriptive enough what I should do.  Example, how will I know what kind of counter name/type to support?  Maybe  a wrapper S3FileIOMetrics interface extend FileIO interface with methods like:
```
 Counter<Long> readBytesCounter() {
   counter(READ_BYTES, Long.class);
}
```
that S3FileIO calls instead, so it's more clear what to implement?

**Comment by szehon-ho** [comment]
Actually thinking on the second point , the current API is probably fine if you are implementing a generic metrics system, but that the first implementation using Hadoop's FileMetric class is not a generic API.  So maybe this could be just be clarified that HadoopFileIOMetrics works only with S3FileIO (as it expects the exact metrics name/type that S3FileIO calls with), and we should not put it in core

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 5]
```diff
@@ -215,6 +215,12 @@
   public static final String S3_CHECKSUM_ENABLED = "s3.checksum-enabled";
   public static final boolean CLIENT_ENABLE_ETAG_CHECK_DEFAULT = false;
 
+  /**
+   * FileIOMetrics implementation class used to report metrics
+   */
+  public static final String S3_FILEIO_METRICS_IMPL = "s3.fileio.metrics-impl";
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:221]: What's the benefit of making this dynamic? I thought this was just an adapter that can be used to record metrics in `FileIO`. Shouldn't that be the same and we just customize what accepts the metrics later?

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  String READ_OPERATIONS = "read.operations";
+  String WRITE_BYTES = "write.bytes";
+  String WRITE_OPERATIONS = "write.operations";
+
+  default void initialize(Map<String,String> properties) {}
+
+  interface Counter<T> {
+    void increment(T amount);
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java:37]: Maybe add `default void increment() { increment(1); }`?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputFile.java [hunk 7]
```diff
@@ -49,7 +47,6 @@ public long getLength() {
 
   @Override
   public SeekableInputStream newStream() {
-    return new S3InputStream(client(), uri(), awsProperties());
+    return new S3InputStream(client(), uri(), awsProperties(), metrics());
   }
-
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputFile.java:54]: Nit: unnecessary whitespace change.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java [hunk 8]
```diff
@@ -46,17 +48,23 @@
...
   }
 
-  S3InputStream(S3Client s3, S3URI location, AwsProperties awsProperties) {
+  S3InputStream(S3Client s3, S3URI location, AwsProperties awsProperties, FileIOMetrics metrics) {
     this.s3 = s3;
     this.location = location;
     this.awsProperties = awsProperties;
 
+    readBytes = metrics.counter(FileIOMetrics.READ_BYTES, Long.class);
+    readOperations = metrics.counter(FileIOMetrics.READ_BYTES, Integer.class);
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java:66]: Should be `this.readOperations` since it sets an instance field.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIOMetrics.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  @Override
+  public <T> Counter<T> counter(String name, Class<T> type) {
+    switch (name) {
+      case READ_BYTES:
+        ValidationException.check(type != Long.class, "'%s' requires Long type", READ_BYTES);
+        return v -> statistics.incrementBytesRead((Long) v);
+      case READ_OPERATIONS:
+        ValidationException.check(type != Integer.class, "'%s' requires Integer type", READ_OPERATIONS);
+        return v -> statistics.incrementReadOps((Integer) v);
+      case WRITE_BYTES:
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIOMetrics.java:62]: Where are the write counters updated?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 10]
```diff
@@ -51,6 +58,7 @@
    * All fields are initialized by calling {@link S3FileIO#initialize(Map)} later.
    */
   public S3FileIO() {
+    metrics = FileIOMetrics.nullMetrics();
```
[the remark this answers is not in the document]
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:61]: Yes, I'll incorporate this.

**Comment by danielcweeks** [comment]
> Yea it looks good to me too, definitely have heard some asks for it! Two overall questions:
> 
> 1. How about the implementation for S3FileOutputStream to track write_bytes, write_records?
> 2. Not sure what is the right way, but if I am implementing a FileIOMetrics for S3FileIO, the FileIOMetrics interface does not seem descriptive enough what I should do.  Example, how will I know what kind of counter name/type to support?  Maybe  a wrapper S3FileIOMetrics interface extend FileIO interface with methods like:
> 
> ```
>  Counter<Long> readBytesCounter() {
>    counter(READ_BYTES, Long.class);
> }
> ```
> 
> that S3FileIO calls instead, so it's more clear what to implement?

@szehon-ho I think we actually want to keep this rather generic, so we don't want specific counters called out since they may vary by engine and implementation, so that could get quite messy.

After talking with @jackye1995, @nastra and @rdblue about this, I think we're going to up-level the Metrics interfaces so they can be reused for capturing metrics for other areas of the codebase as well.  I'll update with the proposed changes and we can follow on with other meter types (e.g. timers) and incorporate them into areas like scan planning in new PRs.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 5 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: I've removed the override properties for now.  We can always add them back when there's a use-case for overriding the default hadoop implementation.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java (hunk 6 above)
Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java]: I've incorporated this, which reads better for single increments.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 11]
```diff
@@ -103,9 +112,23 @@ private S3Client client() {
 
   @Override
   public void initialize(Map<String, String> properties) {
-    this.awsProperties = new AwsProperties(properties);
-    this.awsClientFactory = AwsClientFactories.from(properties);
-    this.s3 = awsClientFactory::s3;
+    awsProperties = new AwsProperties(properties);
+    s3 = AwsClientFactories.from(properties)::s3;
```
[the remark this answers is not in the document]
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java]: Updated

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3InputFile.java (hunk 7 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3InputFile.java:54]: This is just actually fixing an existing whitespace issue.

**Comment by danielcweeks** [comment]
@nastra @szehon-ho @RussellSpitzer @rdblue @jackye1995 @kbendick 

Thanks for all the feedback.  This is an updated version of the PR that includes the changes discussed including:

- Up-leveling the metrics to be more generic so we can use them in other areas like scan planning
- Renaming to `MetricsContext` and placing in the `org.apache.iceberg.metrics` package to avoid confusion with other metrics concepts

Note there is still a `FileIOMetricsContext` class, but it is purely to further scope the properties we want to use as the standard metrics for the FileIO implementations; this include S3FileIO, but GCSFileIO, OSSFileIO, can be follow-ons that reuse the `HadoopMetricsContext` to provide consistent implementations.

For future use cases, I added a `Unit` concept to the counter so we are explicit about what the metric value is when reporting.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 12]
```diff
@@ -113,18 +120,21 @@
...
+    this.createStack = Thread.currentThread().getStackTrace();
 
-    multiPartSize = awsProperties.s3FileIoMultiPartSize();
-    multiPartThresholdSize =  (int) (multiPartSize * awsProperties.s3FileIOMultipartThresholdFactor());
-    stagingDirectory = new File(awsProperties.s3fileIoStagingDirectory());
-    isChecksumEnabled = awsProperties.isS3ChecksumEnabled();
+    this.multiPartSize = awsProperties.s3FileIoMultiPartSize();
+    this.multiPartThresholdSize =  (int) (multiPartSize * awsProperties.s3FileIOMultipartThresholdFactor());
+    this.stagingDirectory = new File(awsProperties.s3fileIoStagingDirectory());
+    this.isChecksumEnabled = awsProperties.isS3ChecksumEnabled();
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:128]: Thanks for fixing these.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopMetricsContext.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * The Hadoop implementation delegates to the Hadoop delegates to the
+   * FileSystem.Statistics implementation and therefore does not require
+   * support for operations like unit() and count() as the counter
+   * values are not directly consumed.
+   *
+   * @param name name of the metric
+   * @param type numeric type of the counter value
+   * @param unit ignored
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopMetricsContext.java:60]: Why include unit if it is ignored? Is this to supply information for other metrics collection libraries that require it?

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+     * @param amount to be incremented
+     */
+    void increment(T amount);
+
+    /**
+     * Reporting count is optional if the counter is reporting externally.
+     *
+     * @return current count if available
+     */
+    default Optional<T> count() {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:69]: It is a little odd that these are defined, but not implemented in the initial use of `MetricsContext`. How do you intend to use these (`count` and `unit`) later on? Since they have default implementations, can we add them later instead of now when we have a use for them?

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+  /**
+   * Get a named counter of a specific type.  Metric implementations may impose
+   * restrictions on what types are supported for specific counters.
+   *
+   * @param name name of the metric
+   * @param type numeric type of the counter value
+   * @param unit the unit designation of the metric
+   * @return a counter implementation
+   */
+  default <T extends Number> Counter<T> counter(String name, Class<T> type, Unit unit) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:87]: I think it would be better to leave this empty so that it is required for implementations. Then put the default noop counter implementation in `nullMetrics`. That signals what people need to implement.

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * Generalized interface for creating telemetry related instances for tracking
+ * operations.  Implementations must take into account usage considerations
+ * like thread safety and serialization.
+ */
+public interface MetricsContext extends Serializable {
+  enum Unit {
+    UNDEFINED("undefined"),
+    BYTES("bytes"),
+    SCALAR("scalar");
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:35]: Rather than "scalar", what about using "operations" as a unit?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 17]
```diff
@@ -104,8 +112,23 @@ private S3Client client() {
...
+    if (s3 == null) {
+      this.s3 = AwsClientFactories.from(properties)::s3;
+    }
+
+    // Report Hadoop metrics if Hadoop is available
+    try {
+      Class<? extends MetricsContext> clazz = DynClasses.builder().impl(DEFAULT_METRICS_IMPL).buildChecked();
+
+      this.metrics = clazz.getDeclaredConstructor().newInstance();
+      metrics.initialize(ImmutableMap.of("fileio.scheme", "s3"));
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:126]: Do you think that other metrics objects will also be loaded dynamically? If not, I'd probably just pass `"s3"` to the constructor:

```java
    this.metrics = DynConstructors.builder()
        .hiddenImpl("org.apache.iceberg.hadoop.HadoopMetricsContext", String.class)
        .buildChecked()
        .newInstance("s3");
```

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 18]
```diff
@@ -104,8 +112,23 @@ private S3Client client() {
...
+    } catch (ClassNotFoundException e) {
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:127]: I'd recommend using `DynConstructors` rather than `DynClasses` because `DynConstructors` will catch and handle `NoClassDefFoundError` as well as `ClassNotFoundException` and will always throw `ClassNotFoundException`. That handles cases where dependencies of the class aren't found, like when Hadoop is missing.

Otherwise, this may leak the `NoClassDefFoundError` and fail when Hadoop is missing.

I'd also update the error message here to `Unable to load metrics class`. And since this is a warning, consider suppressing the exception because we expect it. Exception traces that are expected tend to confuse people looking at logs.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.metrics.MetricsContext;
+
+/**
+ * Extension of MetricsContext for use with FileIO to define standard metrics
+ * that should be reported.
+ */
+public interface FileIOMetricsContext extends MetricsContext {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: I like having a place for the constants, but what is the value of making an identical subinterface of `MetricsContext` rather than using the parent directly?

Also, instead of just names, should we have some metric definition that includes unit (assuming we move forward with units)?

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Generalized interface for creating telemetry related instances for tracking
+ * operations.  Implementations must take into account usage considerations
+ * like thread safety and serialization.
+ */
+public interface MetricsContext extends Serializable {
```
Comment (jackye1995) [inline api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:31]: We should be careful about naming here, we already have a class called `Metrics` for column stats. Is it possible for us to have a somewhat different name?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java (hunk 17 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:126]: I think scheme might need to be configurable, because we know people are using S3FileIO also for non-S3 platforms and have different file schemes.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java (hunk 19 above)
Comment (jackye1995) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: I think the intention is to have `HadoopFIleIOMetricsContext` accessing these, but we are still switching across different metrics in the implementation and we do not allow unsupported ones. In that case, does it make more sense to just have methods for updating each metrics directly?

Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: Yes, these are metrics that are specific to file system operations to distinguish from other operations.  We could collapse them all, but it could get confusing and maybe we don't want to reuse the same names for different usages.

The unit should be defined by the producer, not required by the metric.  For example, one file IO impl could produce bytes while the other could produce kilobytes.  As long as they annotate the unit they're producing correctly, it should be fine.

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java (hunk 16 above)
Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:35]: how about just "count"

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java (hunk 15 above)
Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:87]: The reason I left it was so that if you have an implementation that only requires one (e.g. Counter vs Timer), you don't need to provide a timer implementation.  It would probably be better to throw Unsupported, so I'll change to that.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java (hunk 19 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: I understand coordinating what metrics are required, but why should `FileIO` require a specific `MetricsContext` sub-interface? Why not be able to supply any `MetricsContext` with these metric names?

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java (hunk 16 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:35]: That sounds fine.

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java (hunk 14 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:69]: Is it possible to implement this for Hadoop so we don't need to use `Optional`?

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIOMetrics.java (hunk 9 above)
Comment (danielcweeks) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopFileIOMetrics.java]: I think @szehon-ho already plumbed through metrics on the write side, so we should be good with spark.

File: api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java (hunk 19 above)
Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: `FileIO` does not specifically require a sub-interface.  You can see that S3FileIO only as a handle on a `MetricsContext`.  The only reason for the sub-interface is to have a place to put FileIO specific things, though right now that is only the name of the generic metric names.  These metrics need to be translated by the metrics implementations (in the hadoop case it would be `read.bytes -> statistics.incrementBytesRead()`).

We cannot put the metric names in the HadoopMetricsContext because that would pull in dependencies that would fail class loading when hadoop is not available. 

Alternatively we could just have a an interface/class as the holder for FileIO Metrics, but since this is scoped to usage within the FileIO space, it makes sense to just define a sub-interface to expose the metrics names.

Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: Other implementations of metrics (for example if there is a specific Flink metrics framework) would need to handle these same metric names since that is what the FileIO implementations are producing.

The structure of the interfaces is just to help keep that scoped to FileIO and what is expected to be produce by FileIO implementations.

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java:28]: I think it's a bit confusing to create this interface just for metric names. I'd probably add the metric names as `METRIC_READ_BYTES` and similar on the FileIO interface itself. But this is fairly minor as long as we aren't actually requiring this interface anywhere in code.

File: api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java (hunk 15 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/metrics/MetricsContext.java:87]: Can't implementations throw `UnsupportedOperationException`?

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopMetricsContext.java [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIOMetricsContext;
+
+/**
+ * FileIO Metrics implementation that delegates to Hadoop FileSystem
+ * statistics implementation using the provided scheme.
+ */
+public class HadoopMetricsContext implements FileIOMetricsContext {
+  public static final String SCHEME = "fileio.scheme";
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopMetricsContext.java:36]: How about `io.metrics-scheme`?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good to me once tests are passing.
